### ISSUE TITLE
smooth transitions between switching components

### ISSUE THREAD

**Issue by mgroth0** [rjaros/kvision#105]
I'm creating one of my first websites to display data. 

You can see the WIP page [here](https://www.wolframcloud.com/obj/mjgroth/NoveltyReport/index.html#!/profiles).

Click the numbers on the left and use ctrl + the up and down arrow keys to naviage.

I'm wondering if there is a way to switch these images without the screen flickering and irritating people's eyes.

This may be due to a lack of experience on my part, or it may be an inherent limitation of kvision. Probably the former and if so I apologize but I wasn't sure where else to ask this.

When you switch between subjects, this code is eventually called to make the change:

(this is called within a VPanel that's inside the tab)
```kotlin
//  inspectChildrenRetrieval constructs all the components and returns a List<Widget>
// I was hoping this would reduce the flickering by preloading and preparing the images
val hopefullyPrepped = inspectChildrenRetrieval(obj)

// removes all children from the VPanel. 
// I see at the bottom of this, the function refresh() is called. 
// Perhaps this is bad for my situation? 
// Is there some way to avoid the refresh() here to reduce visual flicker?
removeAll()

// obviously the main flicker happens here, when everything is added. 
// The images seem to appear at the wrong size and then quickly fix themselves. 
// How can I make them appear at the right size and location right away?
addAll(hopefullyPrepped)

```

**Comment by rjaros** [comment]
There is a `protected fun <T> singleRender(block: () -> T): T` method in the `Widget` class. But as you see it's protected, so you would have to create your own class inheriting from `VPanel`. After that, you can try putting your actions inside the `block` lambda. It would look like this:
```
class MyVPanel(...): VPanel(...) {
   fun showContent() {
      singleRender {
         removeAll()
         addAll(hopefullyPrepped)
      }
   }
}
```
See how it's used for instance inside `StateBinding` component: https://github.com/rjaros/kvision/blob/822ee9a8d0efd5bcacacd7dc5c1c35b549ed9c34/src/main/kotlin/pl/treksoft/kvision/state/StateBinding.kt#L38-L50

**Comment by mgroth0** [comment]
Thanks, this looks like it should work but for some reason the images are still not appearing at the correct size in at the moment of rerendering.

My current implementation is:

```kotlin
class MyVPanel : VPanel() {
    fun replaceContentSingleRender(hopefullyPrepped: List<Widget>) {
        singleRender {
            removeAll()
            addAll(hopefullyPrepped)
        }
    }
}

class Inspector<T>(val inspectChildrenRetrieval: (T) -> List<Widget>) {
    var inspectPanel: MyVPanel? = null
    fun inspect(obj: T) {
        inspectPanel?.apply {
            val hopefullyPrepped = inspectChildrenRetrieval(obj)
            replaceContentSingleRender(hopefullyPrepped)
        }
    }

    fun buildInspectorPanel() = MyVPanel().apply {
        inspectPanel = this
    }
}
```
I thought perhaps this issue might be due to me loading images in the wrong way, so I also made some additional changes:
- all Images are now preloaded into a Map 
- responsive is now set to false for all Images

Yet still, even with `singleRender` and pre-loaded images, you can still see some flickering and things not appearing all at once.

**Comment by rjaros** [comment]
KVision doesn't preload images. 

The `<img>` HTML tags are just rendered on the page, and the browser loads the images from the given urls. You can of course set width and height of every image, and the flickering will be even less visible, but unless you implement your own preload mechanism, you will not be able to get rid of it completely.

**Comment by mgroth0** [comment]
Thanks, I've developed an example of the image preloading I'd like to try [here](https://www.w3schools.com/code/tryit.asp?filename=G9RCJSMF03DZ).

Can you recommend the preffered way to implement this kind of preloading in kvision? Would the `imgs` array I've created within the `<script>` tag would even be visible from Kotlin? Or is there a better way I can do this that makes use of the kvision `Image` object?

**Comment by mgroth0** [comment]
I had an idea to test if using an org.w3c.dom.Image class instead of the kivision image class would work. But it seems I can't directly add a org.w3c.dom.Image to a VPanel.

**Comment by rjaros** [comment]
I've come up with this function, which should preload data of all `Image` objects given as a parameter:
```
fun preloadImages(images: List<Image>): Promise<Boolean> {
    val promises = images.map {
        Promise<org.w3c.dom.Image> { resolve, reject ->
            val img = org.w3c.dom.Image()
            img.onload = {
                resolve(img)
            }
            img.onerror = { _, _, _, _, _ ->
                reject(Exception())
            }
            img.onabort = {
                reject(Exception())
            }
            img.src = it.src
        }
    }
    return Promise.all(promises.toTypedArray()).then { true }
}
```
You can use it like this to show the container when all images are ready:
```
val images: List<Image> = ....
hPanel {
    visible = false
    val promise = preloadImages(images)
    addAll(images)
    promise.then { visible = true }
}
```

**Comment by mgroth0** [comment]
I'm wrestling with this now and on my first attempt it doesn't appear to work, and I'm not sure I understood why it would either. `img` has been preloaded yes, but then that object must be garbage collected and I don't understand why that would affect anything when `addAll(images)` is called. I'll keep looking into this in case I'm missing something.

**Comment by mgroth0** [comment]
My best guess is that you intended for the line 
```kotlin
promise.then { visible = true }
```
to block while each `img` loads. And although the `img` isn't used, this would be enough time for each kvision img to load before `visible = true`.

Based on this logic, simply sleeping the thread somehow for a long enough time should achieve the same result. But I tried that: 
```kotlin
val images: List<Image> ...

visible = false

addAll(images)

// (there's no built in way to sleep, 
// but I figured out a quick and dirty way 
// by loading thousands of images). 
val bigList = mutableListOf<Image>()

(0..5000).forEach {
	bigList.addAll(images.toSet())
}

val promise = preloadImages(bigList)
promise.then { visible = true }

```

The result is that the webpage now freezes for a bunch of seconds and then finally displays the content, but exactly like before the images appear to be downloaded on the spot causing them to change size and flicker.

In conclusion, I don't think any actual loading of the added images is happening until `visible = true`?

**Comment by rjaros** [comment]
It definitely works. I've tested inside the pokedex example app, which currently has the same problem as your app - the images are appearing in a random order - https://kvision-pokedex.netlify.com/.

You can't block inside a JS app, because browser environment is single threaded. But you can use async operations using callback and promises - and this function returns a Promise. This promise will be resolved when all images will be loaded by the browser. So any operations inside this callback will be executed when images are ready do show:
```
promise.then { visible = true }
```
You don't need to generate any additional images. It's enough to call `preloadImages` on the original list. It doesn't matter what happens to the `org.w3c.dom.Image` objects created inside, because we only want the browser to download images data.

**Comment by rjaros** [comment]
Compare how the pokedex app works after adding preloading: 
https://kvision-pokedex-preload.netlify.com/

**Comment by mgroth0** [comment]
Sorry I've been so busy, I'm hoping I can get back to working on this problem again. For now I have just a clarifying question. Is kvision-pokedex-preload.netlify.com supposed to be a fixed version of kvision-pokedex.netlify.com? For me, the "preload" version is worse, because the screen flickers more each time you switch pages. I can even see this page counter and arrow buttons, which are normally on the bottom, go to the top for a second in the "preload" version. The non-preload version looks more smooth.

**Comment by rjaros** [comment]
It's fixed when it comes to image preloading - all images are displayed at once. But it does flickers more indeed, and that's why I didn't commit this version to the official example sources. It would require some more work, which is just not worth it, because as you noticed, the non-preload version is good enough :-)

But the idea of preloading is working fine - I think you should be able to use it in your app without problems (your application layout doesn't depend on images sizes).
